lib/ansible/module_utils/facts/hardware/base.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import platform
+
+from ansible.module_utils.basic import get_all_subclasses
+from ansible.module_utils.six import PY3
+
+from ansible.module_utils.facts.facts import Facts
+
+from ansible.module_utils.facts.collector import BaseFactCollector
+
+
+class Hardware(Facts):
+    """
+    This is a generic Hardware subclass of Facts.  This should be further
+    subclassed to implement per platform.  If you subclass this, it
+    should define:
+    - memfree_mb
+    - memtotal_mb
+    - swapfree_mb
+    - swaptotal_mb
+    - processor (a list)
+    - processor_cores
+    - processor_count
+
+    All subclasses MUST define platform.
+    """
+    platform = 'Generic'
+
+    def __new__(cls, *arguments, **keyword):
```

Comment:
Ah, yup, thats pretty much my thinking (as mentioned in https://github.com/ansible/ansible/pull/23012#discussion_r115606711 before I saw this comment).

I think getting rid of the new/metaclass oddness is doable for 2.4 as long as we dont try to avoid the static imports (there is some interesting stuff we could potentially do with a client/module side plugin loader, but def post-2.4...)

Current plan is to add a fact collector 'tags' so the gather_subset (or something more dynamic) can get the platform='Linux' collectors in lieu of the current metaclass/platform matching)